hypervisor/arch/x86/configs/vacpi.c

```diff
@@ -87,6 +87,12 @@ static struct acpi_table_info acpi_table_template[CONFIG_MAX_VM_NUM] = {
 			.address = 0xFEE00000U, /* Local APIC Address */
 			.flags = 0x1U, /* PC-AT Compatibility=1 */
 		},
+		.ioapic = {
+			.header.type = ACPI_MADT_TYPE_IOAPIC,
+			.header.length = sizeof(struct acpi_madt_ioapic),
+			.id = 0x1U,
+			.addr = VIOAPIC_BASE,
+		},
```

Comment:
Do we need IOAPIC for pre-launched VM?

Reply:
I think so for a normal multi-processors system. (And I plan to remove the vPCI.)

Reply:
Whether to remove vPIC is another story. Without strong reason, we won't do so.

Reply:
For a pre-launched VM, we have the vioapic-related stuff without the IOAPIC MADT table, it's very weird. Besides, vUART will transfer from interrupt mode to poll mode if the guest detect interrupt on vCPU1 while we inject ext-interrupt to vCPU0 (PIC bind to BSP). This was gone after add guest IOAPIC support.

Reply:
Yeah we have a confusing state of code now for pre-launched VM. There is no IOAPIC given to it but still it is initialized in hv. Can we stay with polling mode for vUART? If so, we can remove vPIC and vIOAPIC for pre-launched VMs.

Reply:
Yeah right now the state of code is confusing.
And if you plan to remove vPIC, this is definitely needed. If not, the VM console becomes very slow as vUART enters polling mode.

Reply:
Don’t get your pointer. If I didn’t add the ioapic madt table, the vuart may goes to polling mode because of the vuart or pic problem, not what changes in my code. When ioapic madt table was added, this issue disappeared. So what’s the confusing?

Reply:
"Yeah right now the state of code is confusing" -> I was just echoing your statement that today, there is no IO-APIC for pre-launched VM but still hypervisor initializes vIOAPIC for it.
I agree this patch is needed, if vPIC is going to be removed later on.

Reply:
I can't agree with you on this. For a modern OS/processor, IOAPIC is needed no matter  wheter the old-fashioned legacy device  is removed or not.  :-